tt_metal/distributed/mesh_device.hpp

```diff
@@ -117,9 +124,21 @@ class MeshDevice : public std::enable_shared_from_this<MeshDevice> {
     size_t num_cols() const;
     MeshShape shape() const;
 
+    // Reshapes the logical mesh and re-maps the physical devices to the new logical coordinates.
+    // Reshaping Rules:
+    // 1. The old_shape volume must equal the new_shape volume (i.e. number of devices must remain constant)
+    // 2. Line-to-Line Reshaping (when either dimension is 1):
+    //    - Always possible between 1xN and Nx1 shapes (e.g.: 1x8 <-> 8x1
+    // 3. Grid-to-Grid Reshaping:
+    //    - Only possible if the devices can form a connected physical mesh in the new shape
+    //    - Must maintain physical connectivity between adjacent devices
+    // 4. Line-to-Grid Reshaping:
+    //    - Only possible if the physical devices can form a connected physical mesh in the new shape
+    //    - Example: 1x8 -> 2x4 is possible only if physical mesh permits a 2x4 configuration
+    void reshape(const MeshShape& new_shape);
```

Comment:
@cfjchu  If this function should follow specific reshaping rule, can we return with 'bool'?